### PR TITLE
Support setting allowance in swap form

### PR DESCRIPTION
**Pull request by vyorkin**
TODO:
- [x] Add "close" button to allowance dialog

Fixes:
- [x] Fix the case when there is no `currentTransaction` available after changing (refetching) allowance
- [x] Fix missing transaction when resetting allowance to unlimited

Refactoring:
- [x] Get rid of extra `RenderArea`

### CODE REVIEW

File: src/ui/components/address-action/SingleAsset/SingleAsset.tsx [hunk 1]
```diff
@@ -97,18 +95,7 @@ function FungibleAsset({
...
-                <UIText
-                  as={TextLink}
-                  kind="small/accent"
-                  style={{ color: 'var(--primary)' }}
-                  to={allowanceViewHref}
-                >
-                  Edit
-                </UIText>
-              ) : null}
+              <RenderArea name="single-asset-quantity-right" />
```
Comment (vyorkin) [inline src/ui/components/address-action/SingleAsset/SingleAsset.tsx:98]: So we don't have to pass the `allowanceViewHref` down to the `SingleAsset` component anymore

File: src/ui/pages/SwapForm/SwapForm.tsx [hunk 2]
```diff
@@ -519,11 +540,63 @@ export function SwapForm() {
...
+                  requestedAllowanceQuantityBase={UNLIMITED_APPROVAL_AMOUNT}
+                  value={value}
+                  onSubmit={(quantity) => {
+                    setAllowanceQuantityBase(quantity.toString());
+                    allowanceDialogRef.current?.close();
+                  }}
+                  footerRenderArea="custom-allowance-footer"
+                />
+                <Spacer height={20} />
+                <RenderArea name="custom-allowance-footer" />
```
Comment (vyorkin) [inline src/ui/pages/SwapForm/SwapForm.tsx:593]: I don't like that I had to use the `RenderArea` here, alternatives that that I can think of:

* Provide smth like `renderFooter` callback and do this:
```tsx
renderFooter={() => (
  <>
    <Spacer height={20} />
    ...
  </>
)}
```
* ...

File: src/ui/pages/SwapForm/SwapForm.tsx [hunk 3]
```diff
@@ -501,6 +519,9 @@ export function SwapForm() {
         height={innerHeight >= 750 ? '70vh' : '90vh'}
         containerStyle={{ display: 'flex', flexDirection: 'column' }}
         renderWhenOpen={() => {
+          if (!currentTransaction || approveTransactionQuery.isLoading) {
```
Comment (vyorkin) [inline src/ui/pages/SwapForm/SwapForm.tsx:522]: Looks like we should attempt to render `TransactionConfirmationView` if there is no `currentTransaction`.
The issue here is that `currentTransaction` won't be available until `approveTransactioinQuery` is loaded

File: src/ui/pages/SwapForm/SwapForm.tsx (hunk 2 above)
Comment (vyorkin) [reply src/ui/pages/SwapForm/SwapForm.tsx]: Just added an extra `Spacer` to the `CustomAllowanceForm`

File: src/ui/pages/SwapForm/SwapForm.tsx (hunk 3 above)
Comment (vyorkin) [reply src/ui/pages/SwapForm/SwapForm.tsx]: Setting `keepPreviousData: true` kind of helps but there is still an issue with resetting allowance

File: src/ui/pages/SendTransaction/SendTransaction.tsx [hunk 4]
```diff
@@ -277,8 +278,19 @@ function TransactionDefaultView({
           actionTransfers={actionTransfers}
           singleAsset={singleAsset}
           allowanceQuantityBase={allowanceQuantityBase}
-          allowanceViewHref={allowanceViewHref}
         />
+        {allowanceQuantityBase ? (
```
Comment (everdimension) [inline src/ui/pages/SendTransaction/SendTransaction.tsx:282]: I think we also used to check that `action.type` is `approve` before showing this

Comment (vyorkin) [reply src/ui/pages/SendTransaction/SendTransaction.tsx:282]: Good point, thanks! Added an extra check for `action.type`

File: src/ui/pages/SwapForm/SwapForm.tsx [hunk 5]
```diff
@@ -519,11 +535,69 @@ export function SwapForm() {
...
               />
             </ViewLoadingSuspense>
           );
         }}
-      ></BottomSheetDialog>
+      />
+
+      <BottomSheetDialog
+        ref={allowanceDialogRef}
+        key="allowance"
```
Comment (everdimension) [inline src/ui/pages/SwapForm/SwapForm.tsx:549]: what for?

File: src/ui/pages/SwapForm/SwapForm.tsx [hunk 6]
```diff
@@ -519,11 +535,69 @@ export function SwapForm() {
...
+        height="min-content"
+        containerStyle={{ display: 'flex', flexDirection: 'column' }}
```
Comment (everdimension) [inline src/ui/pages/SwapForm/SwapForm.tsx:551]: is this necessary for this view?

File: src/ui/pages/SwapForm/shared/useApproveHandler.ts [hunk 7]
```diff
@@ -84,6 +86,7 @@ export function useApproveHandler({
       return { ...tx, gas: gasAsHex, gasLimit: gasAsHex };
     },
     staleTime: Infinity,
+    keepPreviousData: true,
```
Comment (everdimension) [inline src/ui/pages/SwapForm/shared/useApproveHandler.ts:89]: This is a bit controversial btw, I think it makes it possible to submit stale transaction

File: src/ui/pages/SwapForm/SwapForm.tsx (hunk 5 above)
Comment (vyorkin) [reply src/ui/pages/SwapForm/SwapForm.tsx:549]: I don't think it is necessary here, added for consistency with the `BottomSheetDialog` above, I'll delete this key

File: src/ui/pages/SwapForm/shared/useApproveHandler.ts (hunk 7 above)
Comment (vyorkin) [reply src/ui/pages/SwapForm/shared/useApproveHandler.ts:89]: I'm thinking about blocking the Send (Approve) button until the new/updated `approveTransaction` is constructed.
I mean literally: `<SignTransactionButton ... disabled={... || allowanceIsLoading}`

Comment (vyorkin) [reply src/ui/pages/SwapForm/shared/useApproveHandler.ts:90]: ✅

File: src/ui/pages/SendTransaction/SendTransaction.tsx (hunk 4 above)
Comment (vyorkin) [reply src/ui/pages/SendTransaction/SendTransaction.tsx]: ✅

**Review by everdimension — APPROVED** [review]

File: src/ui/components/address-action/SingleAsset/SingleAsset.tsx [hunk 8]
```diff
@@ -95,7 +96,7 @@ function FungibleAsset({
                 )}{' '}
                 <AssetLink asset={fungible} address={address} />
               </UIText>
-              <RenderArea name="single-asset-quantity-right" />
+              {elementEnd}
```
Comment (vyorkin) [inline src/ui/components/address-action/SingleAsset/SingleAsset.tsx:99]: this looks better indeed

**Pull request merged**